drivers/s3.go

```diff
@@ -392,6 +395,24 @@ func (os *s3Session) DeleteFile(ctx context.Context, name string) error {
 	return err
 }
 
+var wrongRegionRe = regexp.MustCompile("^AuthorizationHeaderMalformed: The authorization header is malformed; the region '([a-zA-Z0-9-]+)' is wrong; expecting '([a-zA-Z0-9-]+)'")
+
+// Given an error, attempt to adjust our configuration to facilitate a succesful retry. Returns true if we should retry.
+func (os *s3Session) recoverError(err error) bool {
+	results := wrongRegionRe.FindStringSubmatch(fmt.Sprint(err))
+	if len(results) == 0 {
+		return false
+	}
+	newRegion := results[2]
+	oldRegion := *os.s3sess.Config.Region
+	// Avoid an infinite loop; if nothing's going to change don't bother
+	if oldRegion == newRegion {
+		return false
+	}
+	os.s3sess.Config.Region = aws.String(newRegion)
```

Comment:
I was slightly worried about a concurrency problem here — two requests fail, one of them retries, one of them sees `oldRegion == newRegion` and gives up — but I think these sessions are only ever used for individual files so I think it's probably okay. If we're worried about it we could just implement a max depth here via a context variable

Reply:
I believe they are only used for individual files on Catalyst, but they are not meant to be used like that. On go-livepeer and task-runner, 1 session is created and multiple files are accessed, possibly in parallel like for recording stuff.

WDYT of:
 - Making the suggested change below to avoid mutating the AWS SDK object
 - On line 355, read the `os.s3sess` into a var and use it consistently in that func, also sending it as arg to this function
 - Adding a lock on the `os`, only for this function
 - After grabbing the lock, this func can check the arg session received is the same in the struct, and if not, consider the error recovered as well (can retry with the new session)